src/v/archival/archival_policy.cc

```diff
@@ -454,28 +536,40 @@ ss::future<upload_candidate> archival_policy::get_next_candidate(
   model::offset begin_inclusive,
   model::offset end_exclusive,
   storage::log log,
-  const storage::offset_translator_state& ot_state) {
+  const storage::offset_translator_state& ot_state,
+  search_for_compacted_segments search_compacted) {
     // NOTE: end_exclusive (which is initialized with LSO) points to the first
     // unstable recordbatch we need to look at the previous batch if needed.
     auto adjusted_lso = end_exclusive - model::offset(1);
-    auto [segment, ntp_conf, forced] = find_segment(
-      begin_inclusive, adjusted_lso, std::move(log), ot_state);
-    if (segment.get() == nullptr || ntp_conf == nullptr) {
+
+    lookup_result result;
+    std::optional<model::offset> last;
+    if (search_compacted) {
+        result = find_compacted_segment(begin_inclusive, std::move(log));
+    } else {
+        result = find_segment(
+          begin_inclusive, adjusted_lso, std::move(log), ot_state);
+        // We need to adjust LSO since it points to the first
+        // recordbatch with uncommitted transactions data
+        if (result.forced) {
+            last = adjusted_lso;
```

Comment:
Can we use this mechanism to make all uploaded compacted segments aligned to the segments in the manifest? This will simplify things quite a bit. Otherwise we might have a situation when there is an overlap between two segments. It might affect recovery since it will compute the size of the partition incorrectly.

Maybe as a first iteration we can adjust the uploaded segment to the committed offset of the segment in the manifest?